plugins/lsp/default.nix

```diff
@@ -1,231 +1,225 @@
+  callSetup = false;
+  hasSettings = false;
 
-      package = lib.mkPackageOption pkgs [
-        "vimPlugins"
-        "nvim-lspconfig"
-      ] { };
+  maintainers = [ lib.maintainers.HeitorAugustoLN ];
 
-      keymaps = {
-        silent = mkOption {
-          type = types.bool;
-          description = "Whether nvim-lsp keymaps should be silent";
-          default = false;
-        };
+  imports = [ ./language-servers ];
 
-        diagnostic = mkOption {
-          type = with types; attrsOf (either str (attrsOf anything));
-          description = "Mappings for `vim.diagnostic.<action>` functions to be added when an LSP is attached.";
-          example = {
-            "<leader>k" = "goto_prev";
-            "<leader>j" = "goto_next";
-          };
-          default = { };
-        };
+  extraOptions = {
+    keymaps = {
+      silent = mkOption {
+        type = types.bool;
+        description = "Whether nvim-lsp keymaps should be silent";
+        default = false;
+      };
 
-        lspBuf = mkOption {
-          type = with types; attrsOf (either str (attrsOf anything));
-          description = "Mappings for `vim.lsp.buf.<action>` functions to be added when an LSP it attached.";
-          example = {
-            "gd" = "definition";
-            "gD" = "references";
-            "gt" = "type_definition";
-            "gi" = "implementation";
-            "K" = "hover";
-          };
-          default = { };
+      diagnostic = mkOption {
+        type = with types; attrsOf (either str (attrsOf anything));
+        description = "Mappings for `vim.diagnostic.<action>` functions to be added when an LSP is attached.";
+        example = {
+          "<leader>k" = "goto_prev";
+          "<leader>j" = "goto_next";
         };
+        default = { };
+      };
 
-        extra = mkOption {
-          type = with types; listOf helpers.keymaps.deprecatedMapOptionSubmodule;
-          apply = map helpers.keymaps.removeDeprecatedMapAttrs;
-          description = ''
-            Extra keymaps to register when an LSP is attached.
-            This can be used to customise LSP behaviour, for example with "telescope" or the "Lspsaga" plugin, as seen in the examples.
-          '';
-          example = [
-            {
-              key = "<leader>lx";
-              action = "<CMD>LspStop<Enter>";
-            }
-            {
-              key = "<leader>ls";
-              action = "<CMD>LspStart<Enter>";
-            }
-            {
-              key = "<leader>lr";
-              action = "<CMD>LspRestart<Enter>";
-            }
-            {
-              key = "gd";
-              action.__raw = "require('telescope.builtin').lsp_definitions";
-            }
-            {
-              key = "K";
-              action = "<CMD>Lspsaga hover_doc<Enter>";
-            }
-          ];
-          default = [ ];
+      lspBuf = mkOption {
+        type = with types; attrsOf (either str (attrsOf anything));
```

Comment:
I'm thinking we can a) use type coercion to reduce the post-processing and b) use our standard `lib.nixvim.keymaps.mapOptionSubmodule` for increased consistency.

We may need to extend `mapOptionSubmodule` to support `key` defaulting to the attr-name though.

I'm thinking something like
```nix
type = coercedTo str (action: {inherit action;}) mapOptionSubmodule;
```
(Although we'd actually need to use the map-option submodule factory to produce a slightly customised type)

Reply:
I like the idea, but I don't know if it would be possible to make it backwards-compatible. But if it is not a problem, then I think this is better.

Reply:
When migrating plugins to mkNeovimPlugin it's acceptable to introduce some breaking changes. You can mark the old option as removed with instructions on how to migrate

Reply:
> You can mark the old option as removed with instructions on how to migrate

This only really applies when the new option has a different name/location.

If the old "removed" option has the same name as the new option that doesn't work.

In this case, I think we can do a migration without breaking changes. But since the option isn't being renamed, it can wait until after this PR.

Reply:
Yeah I meant for options that we move from the top-level to `settings`, but if this option does not move then there's not much we can do